Gradle build does not support `-t` causing test failures and a broken layers example

## Steps to Reproduce

1. cd `examples/layers/`
1. run `flutter build apk`

Expected result: build completes successfully
Actual result:

```
$ flutter build apk
Target file "lib/main.dart" not found.
```

## Flutter Doctor

```
mit-macbookpro2:layers mit$ flutter doctor
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/mit/dev/github/flutter
    • Framework revision cc2ac9a5c0 (61 minutes ago), 2017-02-15 14:05:22
    • Engine revision b64c88766d
    • Tools Dart version 1.22.0-dev.10.6

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.0)
    • Android SDK at /Users/mit/Library/Android/sdk
    • Platform android-25, build-tools 25.0.0
    • ANDROID_HOME = /Users/mit/Library/Android/sdk
    • Java(TM) SE Runtime Environment (build 1.8.0_112-b16)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.2.1, Build version 8C1002
    • ios-deploy 1.9.0

[✓] IntelliJ IDEA Ultimate Edition (version 2016.3.4)
    • Dart plugin version 163.13137
    • Flutter plugin version 0.1.9.1

[✓] Connected devices
    • Android SDK built for x86 • emulator-5554                        • android-x86 • Android 6.0 (API 23) (emulator)
    • Nexus 5X                  • 00ca05b380789730                     • android-arm • Android 7.0 (API 24)
    • iPhone 7 Plus             • A49DCF86-01E1-4B67-AF64-0DD84CFDD495 • ios         • iOS 10.2 (simulator)
```

Using the instructions in the [readme.md](https://github.com/flutter/flutter/blob/master/examples/layers/README.md) don't work either:

```
mit-macbookpro2:layers mit$ flutter run -t widgets/spinning_square.dart -d emulator-5554
Running 'flutter packages get' in layers...           0.4s
Launching widgets/spinning_square.dart on Android SDK built for x86 in debug mode...
Running 'gradlew assembleDebug'...
Target file "lib/main.dart" not found.

FAILURE: Build failed with an exception.

* Where:
Script '/Users/mit/dev/github/mit-flutter/packages/flutter_tools/gradle/flutter.gradle' line: 216

* What went wrong:
Execution failed for task ':app:flutterBuildDebug'.
> Process 'command '/Users/mit/dev/github/mit-flutter/bin/flutter'' finished with non-zero exit value 1
```

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -141,6 +141,13 @@ class FlutterPlugin implements Plugin<Project> {
         if (target == null) {
             target = 'lib/main.dart'
         }
+        if (project.hasProperty('target')) {
+            target = project.property('target')
+        }
+        File targetFile = project.file(Paths.get(project.flutter.source, target).toString())
+        if (!targetFile.exists()) {
+            throw new GradleException("Target file $targetFile not found")
+        }
 
         if (project.tasks.findByName('flutterBuildX86Jar')) {
             project.compileDebugJavaWithJavac.dependsOn project.flutterBuildX86Jar
@@ -184,6 +191,7 @@ class FlutterTask extends DefaultTask {
     String buildMode
     String localEngine
     String localEngineSrcPath
+    @Input
     String targetPath
 
     File sourceDir

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -102,7 +102,7 @@ Future<String> ensureGradle() async {
   return gradle;
 }
 
-Future<Null> buildGradleProject(BuildMode buildMode) async {
+Future<Null> buildGradleProject(BuildMode buildMode, String target) async {
   // Create android/local.properties.
   File localProperties = fs.file('android/local.properties');
   if (!localProperties.existsSync()) {
@@ -130,7 +130,7 @@ Future<Null> buildGradleProject(BuildMode buildMode) async {
     case FlutterPluginVersion.managed:
       // Fall through. Managed plugin builds the same way as plugin v2.
     case FlutterPluginVersion.v2:
-      return buildGradleProjectV2(gradle, buildModeName);
+      return buildGradleProjectV2(gradle, buildModeName, target);
   }
 }
 
@@ -153,7 +153,7 @@ Future<Null> buildGradleProjectV1(String gradle) async {
   printStatus('Built $gradleAppOutV1 (${getSizeAsMB(apkFile.lengthSync())}).');
 }
 
-Future<Null> buildGradleProjectV2(String gradle, String buildModeName) async {
+Future<Null> buildGradleProjectV2(String gradle, String buildModeName, String target) async {
   String assembleTask = "assemble${toTitleCase(buildModeName)}";
 
   // Run 'gradle assemble<BuildMode>'.
@@ -167,6 +167,9 @@ Future<Null> buildGradleProjectV2(String gradle, String buildModeName) async {
     LocalEngineArtifacts localEngineArtifacts = artifacts;
     printTrace('Using local engine: ${localEngineArtifacts.engineOutPath}');
     command.add('-PlocalEngineOut=${localEngineArtifacts.engineOutPath}');
+  }
+  if (target != null) {
+    command.add('-Ptarget=$target');
   }
   command.add(assembleTask);
   int exitcode = await runCommandAndStreamOutput(

--- a/packages/flutter_tools/lib/src/commands/build_apk.dart
+++ b/packages/flutter_tools/lib/src/commands/build_apk.dart
@@ -589,7 +589,7 @@ Future<Null> buildAndroidWithGradle(
     throwToolExit('Try re-installing or updating your Android SDK.');
   }
 
-  return buildGradleProject(buildMode);
+  return buildGradleProject(buildMode, target);
 }
 
 Future<Null> buildApk(

